include/envoy/http/async_client.h

```diff
@@ -50,6 +83,39 @@ class AsyncClient {
     virtual void cancel() PURE;
   };
 
+  /**
+   * An in-flight HTTP stream
+   */
+  class Stream {
+  public:
+    virtual ~Stream() {}
+
+    /***
+     * Send headers to the stream
+     * @param headers supplies the headers to send
+     * @param end_stream supplies whether this is a header only request
+     */
+    virtual void sendHeaders(HeaderMap& headers, bool end_stream) PURE;
+
+    /***
+     * Send data to the stream
+     * @param data supplies the data to send
+     * @param end_stream supplies whether this is the last data
+     */
+    virtual void sendData(Buffer::Instance& data, bool end_stream) PURE;
+
+    /***
+     * Send trailers. This implicitly ends the stream.
+     * @param trailers supplies the trailers to send
+     */
+    virtual void sendTrailers(HeaderMap& trailers) PURE;
+
+    /***
+     * Close the stream.
+     */
+    virtual void close() PURE;
```

Comment:
This is comparable to `Request::cancel()` and its implementation are almost resetting stream and cleanup. I can rename it to reset().

The underlying `Filter` (i.e. router_)'s interface doesn't have reset reason, so I didn't provide that here. Do you think it is needed?